File: mobile_ui/quickchat_ui/lib/features/chat/data/repositories/chat_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:injectable/injectable.dart';
import 'package:quickchat_ui/core/logic/cache_exeption.dart';
import 'package:quickchat_ui/core/logic/error_messages.dart';
import 'package:quickchat_ui/core/logic/result.dart';
import 'package:quickchat_ui/core/logic/server_exeption.dart';
import 'package:quickchat_ui/core/network/network_info.dart';
import 'package:quickchat_ui/features/chat/data/datasources/chat_local_datasource.dart';
import 'package:quickchat_ui/features/chat/data/datasources/chat_remote_datasource.dart';
import 'package:quickchat_ui/features/chat/data/dto/ConversationData.dto.dart';
import 'package:quickchat_ui/features/chat/data/dto/Conversations.dto.dart';
import 'package:quickchat_ui/features/chat/data/dto/ConversationsData.dto.dart';
import 'package:quickchat_ui/features/chat/data/dto/GetConversations.dto.dart';
import 'package:quickchat_ui/features/chat/domain/repositories/chat_repository.dart';

@Environment(Environment.prod)
@LazySingleton(as: ChatRepository)
class ChatRepositoryImpl implements ChatRepository {
  ChatRepositoryImpl({
    required this.networkInfo,
    required this.chatRemoteDatasource,
    required this.chatLocalDatasource,
  });
  final NetworkInfo networkInfo;
  final ChatRemoteDatasource chatRemoteDatasource;
  final ChatLocalDatasource chatLocalDatasource;

  @override
  Stream<Result<ConversationsDto>> connectChatStream() {
    late String accessToken;
    try {
      accessToken = chatLocalDatasource.getToken();
    } catch (e) {
      if (e is CacheException) {
        return Stream<Result<ConversationsDto>>.value(
            Result<ConversationsDto>.fail(error: e.message));
      } else {
        return Stream<Result<ConversationsDto>>.value(
            Result<ConversationsDto>.fail(error: e.toString()));
      }
    }

    final Stream<dynamic> stream =
        chatRemoteDatasource.connectChatStream(accessToken);

    return stream.map((dynamic event) {
      if (event is ConversationsDto) {
        return Result<ConversationsDto>.ok(data: event);
      } else if (event is String) {
        return Result<ConversationsDto>.fail(error: event);
      } else {
        return Result<ConversationsDto>.fail(
            error: 'Unknown error: ${event.toString()}');
      }
    });
  }

  @override
  void disposeChatStream() {
    chatRemoteDatasource.disposeChatStream();
  }

  @override
  Future<Result<ConversationDataDto>> getConversation(
      {required int page}) async {
    if (await networkInfo.isConnected) {
      try {
        final ConversationDataDto loginResult =
            await chatRemoteDatasource.getConversation(page: page);

        return Result<ConversationDataDto>.ok(data: loginResult);
      } on ServerException catch (e) {
        return Result<ConversationDataDto>.fail(error: e.message);
      }
    } else {
      return const Result<ConversationDataDto>.fail(
          error: ERR_NO_INTERNET_CONNECTION);
    }
  }

  @override
  Future<Result<ConversationsDataDto>> getConversations(
      {required GetConversationsDto getConversations}) async {
    if (await networkInfo.isConnected) {
      try {
        final ConversationsDataDto result =
            await chatRemoteDatasource.getConversations(
          conversationsId: getConversations.conversationsId,
          page: getConversations.page,
        );

        return Result<ConversationsDataDto>.ok(data: result);
      } on ServerException catch (e) {
        return Result<ConversationsDataDto>.fail(error: e.message);
      }
    } else {
      return const Result<ConversationsDataDto>.fail(
          error: ERR_NO_INTERNET_CONNECTION);
    }
  }

  @override
  Future<Result<Unit>> sendChat({
    required String recipient,
    required String content,
  }) async {
    if (await networkInfo.isConnected) {
      try {
        final Unit loginResult = await chatRemoteDatasource.sendChat(
          recipient: recipient,
          content: content,
        );

        return Result<Unit>.ok(data: loginResult);
      } on ServerException catch (e) {
        return Result<Unit>.fail(error: e.message);
      }
    } else {
      return const Result<Unit>.fail(error: ERR_NO_INTERNET_CONNECTION);
    }
  }
}

File: web_ui/app/features/chat/data/repositories/chat_repository_impl.ts
import { Result } from "@/app/core/result";
import { ChatRepository } from "../../domain/repositories/chat_repository";
import { ConversationDataDto } from "../dto/ConversationData.dto";
import { ChatRemoteDatasource } from "../datasources/chat_remote_datasource";
import { ConversationsDataDto } from "../dto/ConversationsData.dto";
import { Observable, of, shareReplay } from "rxjs";
import { catchError, finalize, map } from "rxjs/operators";
import { ConversationsDto } from "../dto/Conversations.dto";

export class ChatRepositoryImpl implements ChatRepository {
  constructor(private readonly chatRemoteDatasource: ChatRemoteDatasource) {}
  private sharedStream$: Observable<Result<ConversationsDto>> | null = null;
  connectChatStream(): Observable<Result<ConversationsDto>> {
    let accessToken = localStorage.getItem("access_token");
    if (!accessToken) {
      return of(Result.failure<ConversationsDto>("No access token available"));
    }

    return this.chatRemoteDatasource.connectChatStream(accessToken).pipe(
      map((data) => Result.success<ConversationsDto>(data)),
      catchError((error) => of(Result.failure<ConversationsDto>(error.message)))
    );
  }

  async disposeChatStream(): Promise<Result<void>> {
    try {
      this.chatRemoteDatasource.disposeChatStream();
      return Result.success<void>(undefined);
    } catch (error) {
      return Result.failure<void>(
        error instanceof Error ? error.message : "Unknown error"
      );
    }
  }

  async getConversations(
    conversationsId: string,
    page: number
  ): Promise<Result<ConversationsDataDto>> {
    try {
      const response = await this.chatRemoteDatasource.getConversations(
        conversationsId,
        page
      );
      return Result.success(response);
    } catch (error) {
      return Result.failure(
        error instanceof Error ? error.message : "Unknown error"
      );
    }
  }

  async getConversation(page: number): Promise<Result<ConversationDataDto>> {
    try {
      const response = await this.chatRemoteDatasource.getConversation(page);
      return Result.success(response);
    } catch (error) {
      return Result.failure(
        error instanceof Error ? error.message : "Unknown error"
      );
    }
  }

  async sendMessage(
    recipient: string,
    content: string
  ): Promise<Result<boolean>> {
    try {
      const response = await this.chatRemoteDatasource.sendMessage(
        recipient,
        content
      );
      return Result.success(response);
    } catch (error) {
      return Result.failure(
        error instanceof Error ? error.message : "Unknown error"
      );
    }
  }
}
